src/tablegen.cc

```diff
@@ -467,6 +470,21 @@ Fn_Def *Tablegen::gen_tab() {
   a->add_arg(new Expr::Less(off, new Expr::Fn_Call(new std::string("size"))));
   c.push_back(a);
 
+  if (checkpoint_) {
+    // add std::lock_guard to ensure that the archiving thread
+    // can't read the array while the value is being set
+    Statement::Fn_Call *m = new Statement::Fn_Call(
+                             "std::lock_guard<std::mutex> lock");
+    m->add_arg(new std::string("m"));
+
+    // increase the counter tracking how many cells have been tabulated
+    Statement::Increase *inc_tab_c = new Statement::Increase(
+                                     new std::string("tabulated_vals_counter"));
+
+    c.push_back(m);
```

Comment:
can you move line 484 above 480, such that they are in sync with generated code, i.e. first acquire mutex, then increase

Reply:
Done.